app/models/student/mappers/base.rb

```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+class Student
+  module Mappers
+    class Base
+      attr_reader :payload, :establishment, :year
+
+      def initialize(payload, establishment)
+        @payload = payload
+        @establishment = establishment
+        @year = ENV.fetch("APLYPRO_SCHOOL_YEAR")
+      end
+
+      def parse!
+        classes_with_students.each do |klass, students_attrs|
+          map_students(students_attrs).each do |student|
+            Schooling.find_or_create_by!(classe: klass, student: student)
+          end
+        end
+
+        check_schoolings!
+      end
+
+      def check_schoolings!
+        payload
+          .map { |entry| [entry, Student.find_by(ine: map_student_attributes(entry)[:ine])] }
```

Comment:
Bizarre pour le lecteur d'avoir cette méthode qui n'existe pas dans la classe appelée ici. Là où je bossais avant, mon CTO faisait des : 

```ruby
  def map_student_attributes
    raise "Should be implemented in inherited classes"
  end
```

ça aide à la lecture et ça force le dev qui veut pouvoir hériter de ta classe à implémenter la methode

Reply:
je fais plus trop de contrats d'interface aussi rigides en Ruby, ça casse la nature dynamique du langage en le rapprochant des `Interface` de Java.

https://stackoverflow.com/questions/512466/how-to-implement-an-abstract-class-in-ruby

Note qu'ici `map_student_attributes` peut très bien finir dans `Base`, j'ai failli mais je me suis dit refactor plus tard (puisque l'opération de mapping JSON -> Student.attributes et la même, c'est juste l'origine des flèches qui varient selon FREGATA/SYGNE). Et que le dev qui hérite de ma classe sera obligé d'implémenter cette méthode de toute façon, sinon son code plantera.

Reply:
met juste la méthode sans le raise alors peut-être ?